### PR TITLE
#10133: Migrate update_cache, fill_cache op

### LINKED ISSUE #10133: Move update_cache op

### PR DESCRIPTION
**Pull request by VirdhatchaniKN**
Issue : #10133 

**Work done :**

- Retained both implementations of `update_cache` and `fill_cache`
- `ttnn::kv_cache::fill_cache_for_user_` and `ttnn::kv_cache::update_cache_for_token_` were already present in `ttnn/operations/kv_cache.hpp`  , and now moved to to `ttnn/operations/kv_cache`
- Also migrated `tt_lib.tensor.update_cache` as `ttnn.update_cache` and `tt_lib.tensor.fill_cache` as  `ttnn.fill_cache` at location `ttnn/operations/kv_cache`

**Updates :** 
~Migration completed but facing assertion error with update_cache but fill_cache works fine~ 
`sizeof(::bfloat16)` works

CI : 

- [x] [All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/runs/10414035345) - PASSED
- [x] [[post-commit] C++ tests](https://github.com/tenstorrent/tt-metal/actions/runs/10414038024) - PASSED
- [x] [Device perf regressions and output report](https://github.com/tenstorrent/tt-metal/actions/runs/10414041092) - PASSED
- [x] [[post-commit] models tests](https://github.com/tenstorrent/tt-metal/actions/runs/10414039293) - PASSED
- [ ] [Model perf regressions and output report](https://github.com/tenstorrent/tt-metal/actions/runs/10414036828)
- [x] [(T3K) T3000 model perf tests](https://github.com/tenstorrent/tt-metal/actions/runs/10414042309) - PASSED
- [ ] [(T3K) T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/runs/10414043481)
- [ ] [(T3K) T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/runs/10414045971)
- [ ] [Nightly fast dispatch tests](https://github.com/tenstorrent/tt-metal/actions/runs/10414046338)
- [ ] [(T3K) T3000 frequent tests](https://github.com/tenstorrent/tt-metal/actions/runs/10414046975) - PASSED
- [ ] [(Single-card) Demo tests](https://github.com/tenstorrent/tt-metal/actions/runs/10420867201) - PASSED

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/update_cache/kv_cache.hpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+namespace kv_cache {
+constexpr auto fill_cache_for_user_ = ttnn::register_operation_with_auto_launch_op<
+    "ttnn::kv_cache::fill_cache_for_user_",
+    ttnn::operations::kv_cache::ExecuteFillCache>();
+constexpr auto update_cache_for_token_ = ttnn::register_operation_with_auto_launch_op<
+    "ttnn::kv_cache::update_cache_for_token_",
+    ttnn::operations::kv_cache::ExecuteUpdateCache>();
+
+namespace kv_cache_utils {
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/update_cache/kv_cache.hpp:57]: remove this extra namespace. keep it in ttnn namespace

File: ttnn/cpp/ttnn/operations/update_cache/kv_cache.hpp [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+struct ExecuteStructUpdateCache {
+    static ttnn::Tensor operator()(
+        const ttnn::Tensor& cache,
+        const ttnn::Tensor& input,
+        const uint32_t update_idx,
+        const uint32_t batch_offset,
+        std::optional<const DeviceComputeKernelConfig> compute_kernel_config = std::nullopt);
+};
+
+struct ExecuteStructFillCache {
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/update_cache/kv_cache.hpp:39]: FillCacheOperation

File: ttnn/cpp/ttnn/operations/update_cache/kv_cache.hpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+struct ExecuteUpdateCache {
+    static ttnn::Tensor operator()(
+        const ttnn::Tensor& cache,
+        const ttnn::Tensor& input,
+        const uint32_t update_index,
+        const uint32_t batch_offset,
+        std::optional<const DeviceComputeKernelConfig> compute_kernel_config = std::nullopt);
+};
+
+struct ExecuteStructUpdateCache {
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/update_cache/kv_cache.hpp:30]: UpdateCacheOperation

File: ttnn/cpp/ttnn/operations/update_cache/kv_cache_pybind.cpp [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+            py::arg("batch_idx")});
+}
+
+}  // namespace detail
+
+
+void py_bind_kv_cache(py::module& module) {
+    detail::bind_fill_cache_for_user_(module, ttnn::kv_cache::fill_cache_for_user_);
+    detail::bind_update_cache_for_token_(module, ttnn::kv_cache::update_cache_for_token_);
+    detail::bind_update_cache(module, ttnn::kv_cache::kv_cache_utils::update_cache);
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/update_cache/kv_cache_pybind.cpp:174]: should simply be ttnn::update_cache

File: ttnn/cpp/ttnn/operations/update_cache/kv_cache_pybind.cpp [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    detail::bind_fill_cache(module, ttnn::kv_cache::kv_cache_utils::fill_cache);
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/update_cache/kv_cache_pybind.cpp:175]: should simply be ttnn::fill_cache

File: ttnn/cpp/ttnn/operations/update_cache/kv_cache_pybind.cpp [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+        module,
+        operation,
+        doc,
+        ttnn::pybind_overload_t{
+            [](const update_cache_operation_t& self,
+               const ttnn::Tensor& cache,
+               const ttnn::Tensor& input,
+               const uint32_t update_idx,
+               const uint32_t batch_offset,
+               std::optional<const DeviceComputeKernelConfig> compute_kernel_config = std::nullopt) -> ttnn::Tensor {
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/update_cache/kv_cache_pybind.cpp:120]: dont think std::nullopt is necessary here

File: ttnn/cpp/ttnn/operations/update_cache/kv_cache.hpp (hunk 1 above)
Comment (VirdhatchaniKN) [reply ttnn/cpp/ttnn/operations/update_cache/kv_cache.hpp:57]: It was created inside a separate namespace as you suggested. I'll revert this change

Comment (VirdhatchaniKN) [reply ttnn/cpp/ttnn/operations/update_cache/kv_cache.hpp]: Reverted

File: ttnn/cpp/ttnn/operations/kv_cache/device/kernels/dataflow/reader_fill_cache_interleaved_start_id.cpp [hunk 7]
```diff
@@ -44,4 +44,5 @@ void kernel_main() {
         cb_push_back(cb_id_in0, onetile);
     }
     #endif
+    }
```
Comment (tt-aho) [inline ttnn/cpp/ttnn/operations/kv_cache/device/kernels/dataflow/reader_fill_cache_interleaved_start_id.cpp:47]: This change doesn't look correct

**Review by TT-BrianLiu — APPROVED** [review]

**Review by cglagovichTT — APPROVED** [review]

**Review by ayerofieiev-tt — APPROVED** [review]

**Review by skhorasganiTT — APPROVED** [review]

**Review by eyonland — APPROVED** [review]

**Review by mtairum — APPROVED** [review]

**Review by tt-aho — APPROVED** [review]
Use 2024 for the license header in new files

**Review by johanna-rock-tt — APPROVED** [review]

**Comment by VirdhatchaniKN** [comment]
https://github.com/tenstorrent/tt-metal/actions/runs/10470049067 - passed

**Comment by VirdhatchaniKN** [comment]
https://github.com/tenstorrent/tt-metal/actions/runs/10470055752 - passed

**Comment by VirdhatchaniKN** [comment]
https://github.com/tenstorrent/tt-metal/actions/runs/10470061262 - passed

**Pull request merged**